src/App.js

```diff
@@ -1,13 +1,105 @@
 import React, { Component } from 'react'
+import PropTypes from 'prop-types'
 import { Instructions } from './components'
 import './styles/app.css'
 
+class Counter extends Component {
+  static propTypes = {
+    initialCount: PropTypes.number.isRequired
+  }
+  constructor(props) {
+    super(props)
+    this.state = { count: this.props.initialCount }
+  }
+  componentWillReceiveProps(nextProps) {
+    this.setState({
+      count: nextProps.initialCount
+    })
+  }
+  incClick = () => {
+    this.setState(state => ({
+      count: state.count + 1
+    }))
+  }
+  decClick = () => {
+    this.setState(state => ({
+      count: state.count - 1
+    }))
+  }
+  render() {
+    return (
+      <div>
+        Count state: {this.state.count}
+        <button onClick={this.decClick}>Decrease</button>
+        <button onClick={this.incClick}>Increase</button>
+      </div>
+    )
+  }
+}
+
+class InitialCountForm extends Component {
+  constructor(props) {
+    super(props)
+    this.state = {
+      count: 0,
+      newCount: 0
+    }
+  }
+  handleSubmit(event) {
+    this.setState({
+      count: Number(this.state.newCount),
+      newCount: this.state.newCount
+    })
+    this.props.appCallback(Number(this.state.newCount))
+  }
+  handleForm(event) {
+    this.setState({
+      count: Number(this.state.count),
```

Comment:
you don't have to do this:

if state is:
```
{ x: 1, y: 1}
```

and you call `this.setState({x:2})`, your state becomes:
```
{ x:2, y:1}
```

In other words, `setState` is smart and only overwrites fields in the object that you specify. It doesn't automatically delete fields that you aren't changing. You can rewrite this as:
```javascript
this.setState({newCount: Number(event.target.value)})
```

Some reading: https://reactjs.org/docs/react-component.html#setstate